README.md
# FasterStart
Mod for Factorio

A starter kit with a small modular armor with a small fusion reactor, night vision, exoskeleton, roboport and 25 fusion construction robots.


changelog.txt
---------------------------------------------------------------------------------------------------
Version: 1.0.3
Date: 2020-12-19
  Bugfixes:
    - Fixed an issue where the robots were missing.
---------------------------------------------------------------------------------------------------
Version: 1.0.2
Date: 2020-11-27
  Changes:
    - Code optimized.
---------------------------------------------------------------------------------------------------
Version: 1.0.1
Date: 2020-08-15
  Major Features:
    - Update for Factorio 1.0.
    - Added settings to skip the new freeplay intro.
---------------------------------------------------------------------------------------------------
Version: 0.18.2
Date: 2020-03-14
  Bugfixes:
    - Fixed a sprite issue on low graphics settings.
---------------------------------------------------------------------------------------------------
Version: 0.18.1
Date: 2020-01-24
  Major Features:
    - Update for Factorio 0.18.
---------------------------------------------------------------------------------------------------
Version: 0.17.5
Date: 2019-05-31
  Bugfixes:
    - Fixed a migration issue.
---------------------------------------------------------------------------------------------------
Version: 0.17.4
Date: 2019-05-03
  Changes:
    - Update for factorio 0.17.35.
---------------------------------------------------------------------------------------------------
Version: 0.17.3
Date: 2019-03-10
  Changes:
    - Fixed the version issue. You have to update it manually. I'm sorry :(
---------------------------------------------------------------------------------------------------
Version: 0.17.2
Date: 2019-03-02
  Bugfixes:
    - Fixed a bug that occurred when the player already had armor.
---------------------------------------------------------------------------------------------------
Version: 0.17.1
Date: 2019-03-01
  Major Features:
    - Update for Factorio 0.17.
---------------------------------------------------------------------------------------------------
Version: 0.16.4
Date: 2018-05-27
  Bugfixes:
    - Fixed multiplayer bug.
---------------------------------------------------------------------------------------------------
Version: 0.16.3
Date: 2018-02-19
  Bugfixes:
    - Fixed multiplayer fix.
---------------------------------------------------------------------------------------------------
Version: 0.16.2
Date: 2018-02-26
  Changes:
    - Multiplayer compatibility.
  Bugfixes:
    - Fixed multiplayer bug
---------------------------------------------------------------------------------------------------
Version: 0.16.1
Date: 2017-12-27
  Major Features:
    - Update for Factorio 0.16.
---------------------------------------------------------------------------------------------------
Version: 0.15.0
Date: 2017-07-16
  Major Features:
    - Added a starter kit with a small modular armor with a small fusion reactor, night vision, exoskeleton, roboport and 25 fusion construction robots.

lib.lua
function IsModChanged(data,modName)
	if data 
	 and data.mod_changes 
	 and data.mod_changes[modName] 
	 and data.mod_changes[modName].old_version then 
		return true 
	end
	return false
end

function GetOldVersion(data,modName)
	return FormatVersion(data.mod_changes[modName].old_version)
end

function GetNewVersion(data,modName)
	return FormatVersion(data.mod_changes[modName].new_version)
end

function FormatVersion(version)
	return string.format("%02d.%02d.%02d", string.match(version, "(%d+).(%d+).(%d+)"))
end

function Contains(tab,elem)
	for _,v in pairs(tab) do
		if v == elem then
			return true
		elseif type(v) == "table" then
			if Contains(v,elem) then
				return true
			end
		end
	end
	return false
end

function Remove(tab,elem)
	for i,v in pairs(tab) do
		if v == elem then
			table.remove(tab,i)
			return true
		end
	end
	return false
end


function Count(list)
	local i = 0
	for _ in pairs(list) do
		i = i + 1
	end
	return i
end

data.lua
require("fusion-construction-robot") 

data:extend(
{
	{
        type = "equipment-grid",
		name = "mini-equipment-grid",
		width = 6,
		height = 4,
		equipment_categories = {"armor"}
	},
	
	{
		type = "item",
		name = "fusion-construction-robot",
		icon = "__FasterStart__/graphics/icons/fusion-construction-robot.png",
		icon_size = 64,
		icon_mipmaps = 4,
		subgroup = "logistic-network",
		order = "a[robot]-b[construction-robot]-a[fusion-construction-robot]",
		place_result = "fusion-construction-robot",
		stack_size = 50
	},
	
	{
		type = "item",
		name = "mini-fusion-reactor-equipment",
		icons = 
		{{
			icon = "__base__/graphics/icons/fusion-reactor-equipment.png",
			tint = {r=0.53, g=0.81, b=0.92, a=1}
		}},		
		icon_size = 64,
		icon_mipmaps = 4,
		placed_as_equipment_result = "mini-fusion-reactor-equipment",
		subgroup = "equipment",
		order = "a[energy-source]-b[fusion-reactor]-a[mini-fusion-reactor-equipment]",
		stack_size = 20
	},
	
	{
		type = "armor",
		name = "mini-power-armor",
		icon = "__FasterStart__/graphics/icons/mini-power-armor.png",
		icon_size = 64,
		icon_mipmaps = 4,
		subgroup = "armor",
		order = "c[modular-armor]-a[mini-power-armor]",
		stack_size = 1,
		infinite = true,
		equipment_grid = "mini-equipment-grid",
		inventory_size_bonus = 10
	},
	
	{
		type = "generator-equipment",
		name = "mini-fusion-reactor-equipment",
		sprite =
		{
			filename = "__FasterStart__/graphics/equipment/mini-fusion-reactor-equipment.png",
			width = 65,
			height = 128,
			priority = "medium"
		},
		shape =
		{
			width = 2,
			height = 4,
			type = "full"
		},
		energy_source =
		{
			type = "electric",
			usage_priority = "primary-output"
		},
		power = "250kW",
		categories = {"armor"}
	}
})

control.lua
require("lib")

local modName = "FasterStart"

function Init()
	global = {}
	global.PlayerList = {}
end

function CheckAllPlayers()
	for _,player in pairs(game.players) do
		if player and player.connected then
			GiveArmor(player)		
		end
	end
end

function OnInit()
	Init()
	CheckAllPlayers()
	
	if remote.interfaces["freeplay"] then
		remote.call("freeplay", "set_disable_crashsite", settings.startup['disable-crashsite'].value)
		remote.call("freeplay", "set_skip_intro", settings.startup['skip-intro'].value)
		
		if not settings.startup['disable-crashsite'].value then 
			local ship_items = remote.call("freeplay", "get_ship_items")
			ship_items['fusion-construction-robot'] = (ship_items['fusion-construction-robot'] or 0) + 25
		
			local debris_items = remote.call("freeplay", "get_debris_items")
			debris_items['mini-power-armor'] = (debris_items['mini-power-armor'] or 0) + 1 
			debris_items['mini-fusion-reactor-equipment'] = (debris_items['mini-fusion-reactor-equipment'] or 0) + 1 
			debris_items['night-vision-equipment'] = (debris_items['night-vision-equipment'] or 0) + 1 
			debris_items['exoskeleton-equipment'] = (debris_items['exoskeleton-equipment'] or 0) + 1 
			debris_items['personal-roboport-mk2-equipment'] = (debris_items['personal-roboport-mk2-equipment'] or 0) + 1 
			
			remote.call("freeplay", "set_ship_items", ship_items)
			remote.call("freeplay", "set_debris_items", debris_items)
		end	
	end
end
script.on_init(OnInit)

function OnConfigurationChanged(data)
	if IsModChanged(data,modName) then
		local oldVersion = GetOldVersion(data,modName)
		-- if oldVersion < "00.17.05" or data.mod_changes[modName].old_version == "0.17.201" then
			-- Init()
			-- CheckAllPlayers()
		-- end	
	end
end	
--script.on_configuration_changed(OnConfigurationChanged)

function OnPlayerSpawned(event)
	local player = game.get_player(event.player_index)
	GiveArmor(player)
end	
script.on_event({defines.events.on_player_joined_game,defines.events.on_player_created,},OnPlayerSpawned)

function GiveArmor(player)
	if Contains(global.PlayerList,player) then return end
	if (player.controller_type ~= defines.controllers.character) then return end
	
	local armor_inventory = player.character.get_inventory(defines.inventory.character_armor)
	local player_inventory = player.character.get_inventory(defines.inventory.character_main)
	
	if (armor_inventory.get_item_count("mini-power-armor") > 0 or player_inventory.get_item_count("fusion-construction-robot") > 0) then
		table.insert(global.PlayerList,player)
		return
	end
		
	local contents = armor_inventory.get_contents()
	local armor
		
	if Count(contents) > 0 then 
		player_inventory.insert({name="mini-power-armor"})
		armor = player_inventory.find_item_stack("mini-power-armor")
	else
		armor_inventory.insert({name="mini-power-armor"})
		armor = armor_inventory.find_item_stack("mini-power-armor")
	end
	
	local a_grid = armor.grid
		a_grid.put({name = "mini-fusion-reactor-equipment"})
		a_grid.put({name = "night-vision-equipment"})
		a_grid.put({name = "exoskeleton-equipment"})
		a_grid.put({name = "personal-roboport-mk2-equipment"})
	player_inventory.insert{name="fusion-construction-robot", count = 25}
			
	table.insert(global.PlayerList,player)
end

fusion-construction-robot.lua
local fusion_robot = table.deepcopy(data.raw['construction-robot']['construction-robot'])
fusion_robot.name = "fusion-construction-robot"
fusion_robot.icon = "__FasterStart__/graphics/icons/fusion-construction-robot.png"
fusion_robot.minable.result = "fusion-construction-robot"
fusion_robot. energy_per_tick = "0kJ"
fusion_robot.energy_per_move = "0kJ"
fusion_robot.idle.filename = "__FasterStart__/graphics/entity/fusion-construction-robot.png"
fusion_robot.idle.hr_version.filename = "__FasterStart__/graphics/entity/hr-fusion-construction-robot.png"
fusion_robot.in_motion.filename = "__FasterStart__/graphics/entity/fusion-construction-robot.png"
fusion_robot.in_motion.hr_version.filename = "__FasterStart__/graphics/entity/hr-fusion-construction-robot.png"
fusion_robot.shadow_idle.filename = "__FasterStart__/graphics/entity/fusion-construction-robot-shadow.png"
fusion_robot.shadow_idle.hr_version.filename = "__FasterStart__/graphics/entity/hr-fusion-construction-robot-shadow.png"
fusion_robot.shadow_in_motion.filename = "__FasterStart__/graphics/entity/fusion-construction-robot-shadow.png"
fusion_robot.shadow_in_motion.hr_version.filename = "__FasterStart__/graphics/entity/hr-fusion-construction-robot-shadow.png"
fusion_robot.working.filename = "__FasterStart__/graphics/entity/fusion-construction-robot-working.png"
fusion_robot.working.hr_version.filename = "__FasterStart__/graphics/entity/hr-fusion-construction-robot-working.png"
		
		
data:extend({fusion_robot})

prototypes/faststart.lua
data:extend(
{
	{
		type = "item",
		name = "fusion-construction-robot",
		icon = "__FasterStart__/graphics/icons/fusion-construction-robot.png",
		flags = {"goes-to-main-inventory"},
		subgroup = "logistic-network",
		order = "a[robot]-b[construction-robot]",
		place_result = "fusion-construction-robot",
		stack_size = 25
	},

	{
		type = "item",
		name = "mini-fusion-reactor-equipment",
		icon = "__FasterStart__/graphics/icons/mini-fusion-reactor-equipment.png",
		placed_as_equipment_result = "mini-fusion-reactor-equipment",
		flags = {"goes-to-main-inventory"},
		subgroup = "equipment",
		order = "a[energy-source]-b[fusion-reactor]",
		stack_size = 1
	},
	
	{
        type = "equipment-grid",
		name = "mini-equipment-grid",
		width = 6,
		height = 4,
		equipment_categories = {"armor"}
	},
})

local mini_armor = table.deepcopy(data.raw['armor']['modular-armor'])
mini_armor.name = "mini-power-armor"
mini_armor.icon = "__FasterStart__/graphics/icons/mini-power-armor.png"
mini_armor.resistances = {}
mini_armor.durability = 5000
mini_armor.equipment_grid = "mini-equipment-grid"


local mini_reactor = table.deepcopy(data.raw['generator-equipment']['fusion-reactor-equipment'])
mini_reactor.name = "mini-fusion-reactor-equipment"
mini_reactor.sprite =
    {
      filename = "__FasterStart__/graphics/equipment/mini-fusion-reactor-equipment.png",
      width = 65,
      height = 128,
      priority = "medium"
    }
mini_reactor.shape =
    {
      width = 2,
      height = 4,
      type = "full"
    }
mini_reactor.power = "250kW"


local fcr = table.deepcopy(data.raw['construction-robot']['construction-robot'])
fcr.name = "fusion-construction-robot"
fcr.icon = "__FasterStart__/graphics/icons/fusion-construction-robot.png"
fcr.minable = {hardness = 0.1, mining_time = 0.1, result = "fusion-construction-robot"}
fcr. energy_per_tick = "0kJ"
fcr.energy_per_move = "0kJ"
fcr.idle = 
		{
            filename = "__FasterStart__/graphics/entity/construction-robot-nuclear.png",
            priority = "high",
            line_length = 16,
            width = 32,
            height = 36,
            frame_count = 1,
            shift = {0, -0.15625},
            direction_count = 16
        }
fcr.in_motion = 
		{
            filename = "__FasterStart__/graphics/entity/construction-robot-nuclear.png",
            priority = "high",
            line_length = 16,
            width = 32,
            height = 36,
            frame_count = 1,
            shift = {0, -0.15625},
            direction_count = 16,
            y = 36
        }
fcr.shadow_idle = 
		{
            filename = "__FasterStart__/graphics/entity/construction-robot-nuclear-shadow.png",
            priority = "high",
            line_length = 16,
            width = 50,
            height = 24,
            frame_count = 1,
            shift = {1.09375, 0.59375},
            direction_count = 16
        }
fcr.shadow_in_motion = 
		{
            filename = "__FasterStart__/graphics/entity/construction-robot-nuclear-shadow.png",
            priority = "high",
            line_length = 16,
            width = 50,
            height = 24,
            frame_count = 1,
            shift = {1.09375, 0.59375},
            direction_count = 16
        }
fcr.working = 
		{
            filename = "__FasterStart__/graphics/entity/construction-robot-nuclear-working.png",
            priority = "high",
            line_length = 2,
            width = 28,
            height = 36,
            frame_count = 2,
            shift = {0, -0.15625},
            direction_count = 16,
            animation_speed = 0.3,
        }
fcr.shadow_working = 
		{
            stripes = util.multiplystripes(2, {
                {
                    filename = "__FasterStart__/graphics/entity/construction-robot-nuclear-shadow.png",
                    width_in_frames = 16,
                    height_in_frames = 1,
                }
            }),
            priority = "high",
            width = 50,
            height = 24,
            frame_count = 2,
            shift = {1.09375, 0.59375},
            direction_count = 16
        }
		
		
data:extend({mini_armor,mini_reactor,fcr})
